src/docs/backend/control-silo.mdx

```diff
@@ -0,0 +1,178 @@
+
+### Direct Location Header
+
+When a request is handled by the API Gateway, the response delivered to the client will contain `X-Sentry-Proxy-Url` header which contains the URL that clients should use to avoid the additional latency incurred by the API Gateway.
+
+# Integration Credential Proxy
+
+Integrations can be shared by multiple organizations. For example, a single MS teams workspace can be connected to multiple organizations, and those organizations can live in multiple regions. A few of our integrations leverage refresh tokens and use short lived access tokens. To avoid racy updates when tokens become invalidated and need to be refreshed, we send outbound requests to control silo, which acts as a proxy for the integration service. By funnelling all outbound traffic for integrations that require refresh tokens through control silo we can more easily co-ordinate token replacements.
+
+```mermaid
+sequenceDiagram
+
+actor RegionSilo
+actor ControlSilo
+actor MsTeams
+
+RegionSilo ->> RegionSilo: send issue alert
+RegionSilo ->> ControlSilo : send issue alert to msteams
+ControlSilo ->> ControlSilo : load integration metadata
+ControlSilo ->> ControlSilo : add integration credentials to request
+ControlSilo ->> MsTeams : send notification request with credentials
+MsTeams -->> ControlSilo : token expired!
+ControlSilo ->> MsTeams : refresh token
+MsTeams -->> ControlSilo : response
+ControlSilo ->> ControlSilo : update stored token
+ControlSilo ->> MsTeams : send notification with new token
+MsTeams -->> ControlSilo : response
+ControlSilo -->> RegionSilo : response
+```
+
+The integration proxy is implemented as a class that integrations requiring refresh tokens can sub-class. Currently the following integrations use the integration credential proxy:
+
+- GitLab
+- MSTeams
+- GitHub
+- Vsts (Azure Devops)
+
+### Integration Credential Proxy Headers
+
+The integration credential proxy leverages several specialized headers to perform its role:
+
+- `X-Sentry-Subnet-Organization-Integration` The id of the `OrganizationIntegration` that credentials should be added from.
+- `X-Sentry-Subnet-Base-URL` defines the integration host and base URL that the request should be forwarded to.
+- `X-Sentry-Subnet-Signature` An HMAC based request signature that is verified in control silo.
+- `X-Sentry-Subnet-Path` The path that should be used in the forwarded request.
+- `X-Sentry-Subnet-Keyid` For integrations that use a keyring, the id of the key to use.
+
+# Integration Webhook Forwarding
+
+Many of the 3rd-party services we integrate with only offer a single endpoint that they will submit webhooks to. Because all of our existing webhook traffic is being sent to `[sentry.io](http://sentry.io)` we need to receive webhook requests there. For webhooks that need to be handled in a region silo, we store the webhooks in postgres and then asynchronously deliver captured webhooks to the relevant regions.
```

Comment:
```suggestion
Many of the 3rd-party services we integrate with only offer a single endpoint that they will submit webhooks to. Because all of our existing webhook traffic is being sent to `sentry.io` we need to receive webhook requests there. For webhooks that need to be handled in a region silo, we store the webhooks in postgres and then asynchronously deliver captured webhooks to the relevant regions.
```